### PR TITLE
fix(qrest): release pooled requests before asynchronous processing

### LINKED ISSUE #401: QRest: aggregated FullHttpRequest leaks when a queued request is never answered
Netty's leak detector reports the aggregated HTTP request on a QRest server in production (jPOS-EE 3.0.2-SNAPSHOT, Netty 4.2), roughly once per pod generation at the default `simple` sampling level:

```
LEAK: ByteBuf.release() was not called before it's garbage-collected.
Created at:
  io.netty.buffer.AbstractByteBufAllocator.compositeDirectBuffer(AbstractByteBufAllocator.java:204)
  io.netty.handler.codec.MessageAggregator.decode(MessageAggregator.java:266)
  ...
  io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)
```

`RestSession.channelRead` wraps the request in a `LoggeableHttpRequest`, puts it in the TM `Context` and hands it to `RestServer.queue`; the only release is `SendResponse.releaseRequest` (in `commit` and `abort`). Two paths never reach it:

1. `RestServer.queue` does `sp.out(match.queue(), ctx, 60000L)`. If no TransactionManager takes the context within the lease (saturated TM, or a route mapped to a queue nothing consumes), the entry expires and the retained request is never released.
2. `SendResponse.sendResponse` calls `response.headers()` before entering the `try` whose `finally` releases the request, so a transaction that ends with no response in the context throws an NPE and leaks.

Neither path shows up in the access log, because `RestAccessState` is only completed by `SendResponse`, which is why two sampled records were all the evidence available.

Suggested fix: track the pending request on the channel (an `AttributeKey`) and release it from `RestSession.channelInactive` / `handlerRemoved` while its `refCnt` is still positive, and make `sendResponse` null-safe so the `finally` always runs. Running with `-Dio.netty.leakDetection.level=advanced` would name the last-touching handler if a third path exists.

### PR DESCRIPTION
**Pull request by ar-agt**
## Summary

Closes #401.

- Detach aggregated HTTP requests at the QRest ingress boundary: copy the readable body into a Java byte array, copy headers/trailers/decoder result, and release the original Netty request in a finally block before returning from channelRead.
- Preserve the FullHttpRequest interface and credential-masked LoggeableHttpRequest rendering. Queued snapshots use an unreleasable **unpooled heap** view whose lifetime is managed by GC with the Context; no pooled/direct storage is retained across the Space lease.
- Share one response helper between commit and abort. Its cleanup covers response construction, header processing and sending, and releases response buffers when construction fails or ownership has not been handed to Netty.
- Track access state per Context plus a channel pending set, with synchronized, once-only completion. Complete on the write future, channel close, handler removal or response timeout; failed writes are not recorded as successful responses. Use Context-specific route tagging in Router and Q2Info.
- Add a request-timeout (seconds), defaulting to timeout (300 seconds by default), which closes an unanswered connection and completes pending access state. Successful completion cancels the timer; zero disables the per-request timer. Error responses no longer depend on a non-null exception message or an initialized protocol-version attribute.

## Ownership design / compatibility

This deliberately differs from releasing a channel-held request on disconnect: a TM participant may still be reading that request. There is no generally available QRest transaction-completion hook that makes disconnect-driven release safe without a core/deployment change.

Instead, this self-contained fix pays one bounded body copy (maxContentLength, 512 KiB by default). The original pooled allocation is released deterministically even when the Space entry expires, SendResponse never runs, or queueing throws. The detached heap data remains valid for any participant still holding the Context, including after disconnect. This is not an unreleasable wrapper over pooled memory.

- The 60-second Space lease is unchanged. The response-wait timer is separate and does not claim to detect queue pickup, cancel transactions, or prevent a queued transaction from executing after client disconnection.
- retain/release on QRest-created heap snapshots do not govern their lifetime. Ordinary application-created LoggeableHttpRequest wrappers still delegate reference counting. Explicit copies made by participants follow the rules of their backing buffers.
- The new per-request deadline is not extended by subsequent keep-alive traffic. Configure request-timeout for legitimate long-running requests. It closes the connection without fabricating an HTTP response/status.
- Per-request tracking prevents overwritten audit state, but this PR does not add HTTP/1.1 pipelining response ordering.
- A missing RESPONSE already produces a 404 (or the transaction-failure status); preserve and test that behavior rather than treating it as a null-response leak.

## Tests and verification

Java 26.0.2-amzn and Gradle 9.7.1 selected from .sdkmanrc.

- `./gradlew :modules:qrest:test :modules:qrest:javadoc`
- Final `./gradlew check assemble :modules:qrest:javadoc`: **BUILD SUCCESSFUL**.
- Initial full-run aggregate reports: **391 tests, zero failures/errors, 2 skipped**; QRest: **188 tests, zero failures/errors/skips**.
- PostgreSQL-backed MiniGL/SeqNo and Kafka suites passed in the full run. The final retry reused passing unchanged tasks after fixing a test-clock dependency; no exclusions added.
- Regression tests cover actual HTTP aggregation over pooled direct components and Space lease expiry, detached body/header/trailer integrity, concurrent participant reads during disconnect, multiple pending requests, handler removal, deadline cancellation, competing access completion, queue/write failures, response construction/header failures, and missing-response fallback.
- Netty EmbeddedChannel time is frozen/advanced explicitly for deadline assertions; ownership is verified with reference counts rather than sampled GC leak warnings.
- `git diff --check` passed.

The initial full build emitted two pre-existing non-fatal Javadoc diagnostics in unchanged files: GroovyRequestListener.java (`h1` heading) and qrest-crud/CRUD.java (unexpected closing `p`). The repository configures Javadoc failOnError=false; QRest Javadoc itself passed. No production-load benchmark or heap-copy throughput benchmark was performed. Modules disabled in settings.gradle are outside the standard build.

## Review follow-up (0c4191ad0)

Removed the redundant Context monitor and unused QRestMetrics.requestFailed helper; the production exception-path metrics remain covered by RestSession integration tests. Added ownership comments, clarified timeout behavior, and documented the fixed client-visible HTTP 500 body in the guide and pending ChangeLog entry. Added three CORS ownership regressions and two generic-error-body regressions; removed the obsolete helper-only unit test.

`./gradlew :modules:qrest:check :modules:qrest:assemble :modules:qrest:javadoc` passed with Java 26.0.2-amzn / Gradle 9.7.1: **192 QRest tests, zero failures/errors/skips**. `git diff --check` passed. The unchanged non-QRest suites were not rerun for this review-only follow-up; the full-run results above refer to the initial implementation.

### CODE REVIEW

File: modules/qrest/src/main/java/org/jpos/qrest/RestSession.java [hunk 1]
```diff
@@ -87,15 +95,21 @@ public void channelRead(ChannelHandlerContext ch, Object msg) throws Exception {
...
-            captureRequest(ch, request);
-            Context ctx = new Context();
-            ctx.put(Constants.SESSION, ch);
-            ctx.put(Constants.REQUEST, new LoggeableHttpRequest(request, maskedHeaders));
-            ch.channel().attr(httpVersion).set(request.protocolVersion());
-
-            if (contentKey != null)
-                ctx.put(contentKey, request.content().toString(CharsetUtil.UTF_8));
-            server.queue(request, ctx);
+            try {
```
Comment (ar) [inline modules/qrest/src/main/java/org/jpos/qrest/RestSession.java:100]: The OPTIONS/CORS branch above stays outside this try/finally. I checked Netty 4.2.15's `CorsHandler`: preflight and forbidden both release the request before responding, and the pass-through goes to the pipeline tail, which releases unhandled messages, so there is no third leak path there. Nothing pins it though; a one-line comment on the branch, or a test asserting refCnt 0 after a preflight, would keep the next reader from re-deriving it.

File: modules/qrest/src/main/java/org/jpos/qrest/RestSession.java [hunk 2]
```diff
@@ -120,19 +134,21 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
...
+                if (!state.completed && state.status == null)
+                    state.status = HttpResponseStatus.INTERNAL_SERVER_ERROR.code();
+            }
+        }
 
         ctx.writeAndFlush(new DefaultFullHttpResponse(
           version,
           HttpResponseStatus.INTERNAL_SERVER_ERROR,
-          copiedBuffer(cause.getMessage().getBytes())
+          copiedBuffer("Internal Server Error", CharsetUtil.UTF_8)
```
Comment (ar) [inline modules/qrest/src/main/java/org/jpos/qrest/RestSession.java:153]: Good change: the old body echoed the exception message to the client and threw when the message was null. It is client-visible though, so it belongs in the changelog.

File: modules/qrest/src/main/java/org/jpos/qrest/RestSession.java [hunk 3]
```diff
@@ -58,6 +63,8 @@ public class RestSession extends ChannelInboundHandlerAdapter {
 
     RestSession(RestServer server) {
         this.server = server;
+        requestTimeout = server.getConfiguration().getInt("request-timeout",
```
Comment (ar) [inline modules/qrest/src/main/java/org/jpos/qrest/RestSession.java:66]: With the default equal to `timeout`, a request whose Space lease expires at 60 seconds keeps its connection open until 300 seconds and then closes with no response. The docs say so honestly, and a 503 at lease expiry is out of scope since the Space cannot signal it; just flagging that operators will see the idle timer and this deadline coincide by default.

File: modules/qrest/src/main/java/org/jpos/qrest/SendResponse.java [hunk 4]
```diff
@@ -56,60 +54,62 @@ public int prepare(long id, Serializable context) {
...
-            headers.set(HttpHeaderNames.CONTENT_LENGTH, responseBytes);
-            captureResponse(ch, response, responseBytes);
-            ChannelFuture cf = ch.writeAndFlush(response);
-
-            if (!keepAlive)
-                cf.addListener(ChannelFutureListener.CLOSE);
-        } finally {
-            releaseRequest(ctx, request);
+    private void respond(Context ctx) {
+        synchronized (ctx) {
```
Comment (ar) [inline modules/qrest/src/main/java/org/jpos/qrest/SendResponse.java]: Nit: nothing else locks on the Context, and commit and abort are never both invoked for one transaction, so this monitor has no counterpart. Harmless, but it reads as if it guards a race that does not exist.

File: modules/qrest/src/main/java/org/jpos/qrest/Constants.java [hunk 5]
```diff
@@ -23,6 +23,7 @@
 public enum Constants implements Inhibit {
     SESSION,
     REQUEST,
+    ACCESS_STATE,
```
Comment (ar) [inline modules/qrest/src/main/java/org/jpos/qrest/Constants.java:27]: Worth a one-line comment: this key stays out of the TransactionManager's Context dump only because the enum implements `Inhibit` and `Context.dumpEntry` skips `Inhibit` keys. Anyone who later moves the key to a plain string constant would start dumping the access state.

**Review by ar — APPROVED** [review]
Approving. This fixes both paths from #401, and the ingress copy is a sounder mechanism than the disconnect-driven release I suggested there: releasing on `channelInactive` would free the buffer under a participant that is still reading it, while detaching in `channelRead` ends the pooled buffer's lifetime deterministically on the event loop, bounded by `maxContentLength`. The lease-expiry test over a real `HttpObjectAggregator` with pooled direct chunks, and the missing-response / construction-failure / header-failure tests on `SendResponse`, cover exactly the two scenarios in the report, and the null-status access record on disconnect or deadline fixes the silence that made the original incident hard to diagnose.

One note that has no line in the diff: `QRestMetrics.requestFailed` is no longer called from production code. The 500 still lands, since `exceptionCaught` stamps it and the close completes the state through `requestCompleted`, so either delete the method or note in its Javadoc that it is now test-only.

The inline comments are nits and changelog reminders, none blocking.

File: modules/qrest/src/main/java/org/jpos/qrest/RestSession.java (hunk 1 above)
Comment (ar-agt) [reply modules/qrest/src/main/java/org/jpos/qrest/RestSession.java:100]: Addressed in 0c4191ad0: added the ownership comment and three regressions for preflight, forbidden, and downstream pass-through. Each asserts the pooled request reaches refCnt 0 without entering the TM queue. The pass-through test uses an explicit consuming handler because EmbeddedChannel captures unhandled messages instead of behaving like the normal releasing tail.

File: modules/qrest/src/main/java/org/jpos/qrest/RestSession.java (hunk 2 above)
Comment (ar-agt) [reply modules/qrest/src/main/java/org/jpos/qrest/RestSession.java:153]: Addressed in 0c4191ad0: documented the fixed Internal Server Error body and null-message behavior in the QRest guide, with regressions for both null and non-null exception messages. Also added a ChangeLog entry explicitly marked pending merge of PR402: https://github.com/jpos/jPOS-EE/wiki/ChangeLog (wiki commit 1339999).

File: modules/qrest/src/main/java/org/jpos/qrest/RestSession.java (hunk 3 above)
Comment (ar-agt) [reply modules/qrest/src/main/java/org/jpos/qrest/RestSession.java:66]: Agreed; no timeout/default behavior change. The guide now explicitly notes that an expired 60-second queue entry may wait until approximately 300 seconds before connection closure, with the reader-idle and response timers due at roughly the same time on an otherwise idle connection. It also states neither mechanism detects lease expiry or synthesizes a 503.

File: modules/qrest/src/main/java/org/jpos/qrest/SendResponse.java (hunk 4 above)
Comment (ar-agt) [reply modules/qrest/src/main/java/org/jpos/qrest/SendResponse.java]: Removed the Context monitor in 0c4191ad0. The actual cross-thread coordination remains on the per-request RestAccessState; its concurrent completion regression still passes.

File: modules/qrest/src/main/java/org/jpos/qrest/Constants.java (hunk 5 above)
Comment (ar-agt) [reply modules/qrest/src/main/java/org/jpos/qrest/Constants.java:27]: Added the requested comment on ACCESS_STATE in 0c4191ad0: it must remain an Inhibit key so moving it to a plain String does not expose access state in Context dumps.

**Comment by ar-agt** [comment]
@ar Addressed the five inline notes in 0c4191ad0 and replied to each thread. Also removed the now-unused QRestMetrics.requestFailed method and its helper-only test; the real exception/close metrics path remains tested. QRest check, assembly and Javadoc passed: **192 tests, zero failures/errors/skips**. Timeout defaults are unchanged, with the operator behavior clarified in the guide. The client-visible error-body change is in the wiki ChangeLog, explicitly marked pending merge. No merge performed.

File: modules/qrest/src/main/java/org/jpos/qrest/SendResponse.java [hunk 6]
```diff
@@ -56,60 +54,60 @@ public int prepare(long id, Serializable context) {
...
+            if (ch != null)
+                ch.close();
+            throw e;
         } finally {
-            releaseRequest(ctx, request);
+            try {
+                if (!handedOff && response != null)
+                    ReferenceCountUtil.release(response);
+            } finally {
+                releaseRequest(ctx, request);
```
Comment (ar) [inline modules/qrest/src/main/java/org/jpos/qrest/SendResponse.java:106]: Nit, not blocking: the first version nested this in its own `finally` so the request was released even if `release(response)` threw. With the heap snapshot that no longer leaks anything, so flattening is fine; just noting the guarantee got weaker for application code that puts a plain reference-counted request in the Context.

**Review by ar — COMMENTED** [review]
0c4191ad0 addresses all five notes; the CORS regressions and the two exception-body tests pin exactly what I asked for, and removing `requestFailed` outright is fine since the new tests cover the 500-and-completed path end to end. My approval stands.

One process note: the wiki ChangeLog was edited ahead of the merge, with the PR 402 entry marked pending. I would rather changelog entries wait until the PR is actually merged, so the wiki never describes behavior a release does not have. Please leave it as is for now and, going forward, add the entry as part of the merge rather than before it.

File: modules/qrest/src/main/java/org/jpos/qrest/SendResponse.java (hunk 6 above)
Comment (ar-agt) [reply modules/qrest/src/main/java/org/jpos/qrest/SendResponse.java:106]: Checked 0c4191ad0: the nested try/finally is still present at SendResponse.java lines 101–107. The Context monitor removal only dedented this block; releaseRequest(ctx, request) remains in the inner finally, so it still runs if ReferenceCountUtil.release(response) throws, including for plain reference-counted requests. No code change needed.

https://github.com/jpos/jPOS-EE/blob/0c4191ad07c774c51f22ef43f3ae6fdcc5ef6649/modules/qrest/src/main/java/org/jpos/qrest/SendResponse.java#L101-L107

**Comment by ar-agt** [comment]
@ar Acknowledged the ChangeLog process note: I will leave the current pending entry unchanged as requested, and publish future wiki ChangeLog entries only after their PRs are actually merged. Your approval is noted; no merge performed. The cleanup follow-up is answered inline: the nested finally is still intact in the current head.
